### PR TITLE
Add Nitrokey WebCrypt

### PR DESCRIPTION
**Pull request by szszszsz**
Add Nitrokey WebCrypt application

For now to be available only in the testing releases.

Fixes https://github.com/Nitrokey/nitrokey-3-firmware/issues/273
Connected: https://github.com/Nitrokey/nitrokey-webcrypt-rust/issues/4

Can be tested against:
- https://webcrypt.nitrokey.com/ (tested with Chrome; `Reset and Login -> sign and decrypt`)
- `v0.2-28-g610f42f` / https://github.com/Nitrokey/nitrokey-webcrypt-tests/tree/devel (`make FIDO2`)

Built with:
```
$  make flash-develop EXTRA_FEATURES=apps/test,no-buttons
```

-----------

Edit: NK3 AM test results:
- [report.zip](https://github.com/Nitrokey/nitrokey-3-firmware/files/11918773/report.zip)

### CODE REVIEW

**Comment by szszszsz** [comment]
Flashes, runs and status command works. Not all commands are working though - working on that.

File: components/apps/src/lib.rs [hunk 1]
```diff
@@ -348,6 +357,26 @@ impl<R: Runner> App<R> for FidoApp<R> {
...
+    const CLIENT_ID: &'static str = "webcrypt";
+
+    type Data = ();
+
+    fn with_client(_runner: &R, trussed: Client<R>, _: ()) -> Self {
+        Self::new(trussed)
+    }
+    fn backends(runner: &R) -> &'static [BackendId<Backend>] {
+        const BACKENDS_WEBCRYPT: &[BackendId<Backend>] = &[
+            BackendId::Custom(Backend::SoftwareRsa),
```
Comment (robin-nitrokey) [inline components/apps/src/lib.rs:371]: Does it really need RSA?  Can’t find it in the app source.  If it does, it should also activate the `backend-rsa` feature.

**Review by robin-nitrokey — CHANGES_REQUESTED** [review]

**Comment by sosthene-nitrokey** [comment]
There will probably need to be a bit of rebasing to be done to make it compatible #272

File: components/apps/src/lib.rs (hunk 1 above)
Comment (szszszsz) [reply components/apps/src/lib.rs:371]: Good spot! Will check that one out.
Yes, RSA is used, and tested against test suite, though still WIP (not merged to main branch, and here the WIP branch was used directly).

**Comment by szszszsz** [comment]
1. Moved to cancel-supporting update. Builds. 
2. Currently can crash due to stack overflow on some Webcrypt calls during deserialization, especially when used with PIV. Should be safe to use otherwise.
3. Uses internal filesystem (to be corrected).
4. Can't be merged yet, as it uses side branches with new features.
```
# Required for inject-any-key and hmacsha256p256
trussed = { git = "https://github.com/Nitrokey/trussed", branch = "webcrypt-devel-0.1.0-nk12" }
# Enable this for FIDO2 transport support
ctaphid-dispatch = { git = "https://github.com/Nitrokey/ctaphid-dispatch.git", branch="multiple_apps" }
```

File: components/apps/Cargo.toml [hunk 2]
```diff
@@ -34,11 +35,12 @@ default = [
     "opcard",
     "trussed/clients-4",
 ]
-test = ["piv-authenticator", "trussed/clients-5"]
+test = ["piv-authenticator", "webcrypt", "trussed/clients-6"]
 provisioner = ["provisioner-app", "trussed/clients-5"]
 
 # apps
 secrets-app = ["dep:secrets-app", "backend-auth"]
+webcrypt = ["dep:webcrypt", "backend-auth"]
```
Comment (robin-nitrokey) [inline components/apps/Cargo.toml:43]: Should also require `backend-rsa`.

Comment (szszszsz) [reply components/apps/Cargo.toml]: Weird. It got eaten during rebase I guess. Good you have checked that one again.
Corrected.

**Review by robin-nitrokey — COMMENTED** [review]
Looks good so far.  Regarding the integration, I think only the PRs and eventually tags for the Git dependencies are missing.

**Comment by szszszsz** [comment]
1. Does not crash anymore on NK3AM with the recent Webcrypt release. Works with https://webcrypt.nitrokey.com/ as well.
2. Still uses internal filesystem, but that can be handled in the next PR.
3. No side branches are used anymore, except for the `trussed-staging` backend and Webcrypt itself. Both will be corrected in the next steps.
4. The new extension added to the `trussed-staging` needs to be activated by feature, so it should be transparent for non-test build.

**Comment by szszszsz** [comment]
Just BTW, for some reason flexiber patch is not used:
```
warning: Patch `flexiber v0.1.1 (https://github.com/Nitrokey/flexiber?tag=0.1.1.nitrokey#26f2c804)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
```

**Comment by szszszsz** [comment]
I've just noticed that it breaks on the "fat" LTO setting. Can we use "thin" for now?

**Comment by robin-nitrokey** [comment]
> Just BTW, for some reason flexiber patch is not used:

We use flexiber 0.1.0 but your patch points to 0.1.1.  In order to use it, you have to run `cargo update -p flexiber`.

**Comment by szszszsz** [comment]
1. Moved to EFS
2. Still crashes with `LTO=true`, even with buffer and max field width to 600 bytes. I have added `#[inline(never)]` in multiple places, but this did not solve it.

**Review by daringer — APPROVED** [review]

**Review by daringer — APPROVED** [review]

**Comment by szszszsz** [comment]
Will be merged soon after a proper WebCrypt release is done.

**Comment by szszszsz** [comment]
All ticks green, merging

**Pull request merged**